Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm

```diff
@@ -249,7 +249,42 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
         websiteDataStoreDirectory.get() = [webkitDirectory URLByAppendingPathComponent:@"WebsiteDataStore" isDirectory:YES];
     });
 
-    return [websiteDataStoreDirectory.get() URLByAppendingPathComponent:identifier isDirectory:YES].absoluteURL.path;
+    return websiteDataStoreDirectory.get().get().absoluteURL.path;
+}
+
+void WebsiteDataStore::fetchAllDataStoreIdentifiers(CompletionHandler<void(Vector<UUID>&&)>&& completionHandler)
+{
+    Vector<UUID> identifiers;
+    for (auto identifierString : FileSystem::listDirectory(defaultWebsiteDataStoreRootDirectory())) {
+        if (auto identifier = UUID::parse(identifierString))
+            identifiers.append(*identifier);
+    }
+
+    completionHandler(WTFMove(identifiers));
+}
+
+void WebsiteDataStore::deleteDataStoreWithIdentifier(const UUID& identifier, CompletionHandler<void(const String&)>&& completionHandler)
+{
+    if (!identifier)
+        return completionHandler("Identifier is invalid"_s);
+
+    bool isIdentifierInUse = false;
+    forEachWebsiteDataStore([&](auto& dataStore) {
+        if (auto storeIdentifier = dataStore.identifier(); storeIdentifier == identifier)
+            isIdentifierInUse = true;
+    });
+    if (isIdentifierInUse)
+        return completionHandler("WebsiteDataStore with this identifier is in use"_s);
```

Comment:
I think it would be better in this case just to go ahead with the delete operation. It's not in a programmer's control to ensure that the last reference to a reference counted object has been dropped. And the delete will do what the programmer intended, so long as the programmer does not do new things to trigger storage after the delete.

Reply:
(I'm expecting that removing a data store and then performing a storage operation on it would have the same effect as removing a data store and then invoking the dataStoreWithIdentifier: API with the same identifier and then performing a storage operation on the new data store -- namely, it would re-create that data store identifier on disk.)

Reply:
Sure, updated.